src/frontend/src/optimizer/rule/pull_up_correlated_predicate_agg_rule.rs

```diff
@@ -153,55 +153,32 @@ impl Rule for PullUpCorrelatedPredicateAggRule {
-        // When group by is empty, count would return 0 instead of null.
-        // Unless we can prove that the row generated by the empty group could be eliminated by the top filter.
-        //
-        // A typical example is the count is generate by avg, because avg = sum / count. In this case, when input is empty,
-        // sum is null, so avg is null. And null-rejected expression will be false, so we can still apply this rule and we don't need to generate a 0 value for count.
-        let count_exists = agg_calls
-            .iter()
-            .any(|agg_call| matches!(agg_call.agg_type, AggType::Builtin(PbAggKind::Count)));
-
-        if count_exists {
-            // When group input is empty, not count agg would return null.
+        // If the `group by + proj` doesn't return null for empty input, we can't apply this rule.
+        {
+            // When group input is empty, if the agg is not `count`, it would return null.
             let null_agg_pos = agg_calls
                 .iter()
                 .positions(|agg_call| {
                     !matches!(agg_call.agg_type, AggType::Builtin(PbAggKind::Count))
                 })
                 .collect_vec();
 
-            // If no null agg, bail out.
+            // We don't have null args, so the output will nevert be null. Bail out.
             if null_agg_pos.is_empty() {
                 return None;
             }
 
-            // Try to prove that the expression is null-rejected by the top filter when not-count-aggs are null.
             let mut agg_null_bitset = FixedBitSet::with_capacity(agg.base.schema().len());
             for pos in null_agg_pos {
                 agg_null_bitset.insert(pos);
             }
 
-            // Shift the top project expressions to the right by apply_left schema len, because it is used to check null-rejected by the top filter.
-            let apply_left_schema = apply_left.schema().len();
-            let mut top_proj_null_bitset =
-                FixedBitSet::with_capacity(top_project.base.schema().len() + apply_left_schema);
-            for (i, expr) in top_proj_exprs.iter().enumerate() {
-                if Strong::is_null(expr, agg_null_bitset.clone()) {
-                    top_proj_null_bitset.insert(i + apply_left_schema);
+            // Only all expr in top_proj_exprs are null, we can apply this rule, otherwise, bail out.
+            for expr in &top_proj_exprs {
+                if !Strong::is_null(expr, agg_null_bitset.clone()) {
+                    return None;
                 }
             }
-
-            // Only all expr in conjunctions are null, we can apply this rule, otherwise, bail out.
-            if top_filter
-                .predicate()
-                .conjunctions
-                .iter()
-                .any(|expr| !Strong::is_null(expr, top_proj_null_bitset.clone()))
-            {
-                return None;
-            }
```

Comment:
Why this logic needed to be changed?

Reply:
I don't get why previously we check top_filter. I think we should check the output exprs instead

Reply:
Checking top_filter to ensure the null agg row generated by left outer join finally could be eliminated by the filter. If it could be eliminated, then we can allow the transformation, even if we meet `count`, `array`, and other aggs will generate a not null output when input is empty.

Reply:
Oh, I think the condition can be:
`top_filter is null OR all of top_proj are null`

Reply:
Yes, you can construct a planner test case I think.